cmd/syncthing/generate/generate.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/fs"
+	"github.com/syncthing/syncthing/lib/locations"
+	"github.com/syncthing/syncthing/lib/osutil"
+	"github.com/syncthing/syncthing/lib/protocol"
+	"github.com/syncthing/syncthing/lib/syncthing"
+)
+
+type CLI struct {
+	cmdutil.CommonOptions
+	GUIUser     string `placeholder:"STRING" help:"Override GUI authentication user name"`
+	GUIPassword string `placeholder:"STRING" help:"Override GUI authentication password (use - to read from standard input)"`
+}
+
+func (c *CLI) Run() error {
+	log.SetFlags(0)
+
+	if c.HideConsole {
+		osutil.HideConsole()
+	}
+
+	if c.HomeDir != "" {
+		if c.ConfDir != "" {
+			return fmt.Errorf("--home must not be used together with --config")
+		}
+		c.ConfDir = c.HomeDir
+	}
+	if c.ConfDir == "" {
+		c.ConfDir = locations.GetBaseDir(locations.ConfigBaseDir)
+	}
+
+	// Support reading the password from a pipe or similar
+	if c.GUIPassword == "-" {
+		reader := bufio.NewReader(os.Stdin)
+		password, _, err := reader.ReadLine()
+		if err != nil {
+			return fmt.Errorf("Failed reading GUI password: %w", err)
+		}
+		c.GUIPassword = string(password)
+	}
+
+	if err := Generate(c.ConfDir, c.GUIUser, c.GUIPassword, c.NoDefaultFolder); err != nil {
+		return errors.Wrap(err, "Failed to generate config and keys")
+	}
+	return nil
+}
+
+func Generate(confDir, guiUser, guiPassword string, noDefaultFolder bool) error {
+	dir, err := fs.ExpandTilde(confDir)
+	if err != nil {
+		return err
+	}
+
+	if err := syncthing.EnsureDir(dir, 0700); err != nil {
+		return err
+	}
+	locations.SetBaseDir(locations.ConfigBaseDir, dir)
+
+	var myID protocol.DeviceID
+	certFile, keyFile := locations.Get(locations.CertFile), locations.Get(locations.KeyFile)
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err == nil {
+		log.Println("WARNING: Key exists; will not overwrite.")
+	} else {
+		cert, err = syncthing.GenerateCertificate(certFile, keyFile)
+		if err != nil {
+			return errors.Wrap(err, "create certificate")
+		}
+	}
+	myID = protocol.NewDeviceID(cert.Certificate[0])
+	log.Println("Device ID:", myID)
+
+	cfgFile := locations.Get(locations.ConfigFile)
+	var cfg config.Wrapper
+	if _, err := os.Stat(cfgFile); err == nil {
+		if guiUser == "" && guiPassword == "" {
+			log.Println("WARNING: Config exists; will not overwrite.")
+			return nil
+		}
+
+		if cfg, _, err = config.Load(cfgFile, myID, events.NoopLogger); err != nil {
+			return errors.Wrap(err, "load config")
+		}
+	} else {
+		if cfg, err = syncthing.DefaultConfig(cfgFile, myID, events.NoopLogger, noDefaultFolder); err != nil {
+			return errors.Wrap(err, "create config")
+		}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go cfg.Serve(ctx)
+	defer cancel()
+
+	guiCfg := cfg.GUI()
+	waiter, err := cfg.Modify(func(cfg *config.Configuration) {
+		if changed := overrideGUIAuthentication(&guiCfg, guiUser, guiPassword); changed {
+			cfg.GUI = guiCfg
```

Comment:
This assignment seems a no-op since overrideGUIAuthentication took a pointer a changed the object itself

Reply:
Hm. I believe the Wrapper's `GUI()` method returns a copy though, not the original `GUIConfiguration` object?  That's why we take its address here.  The same pattern can be found in `lib/api/confighandler.go` for example.

Reply:
I believe there's just no need for the `guiCfg` object assigned outside of `Modify`. You could just pass `&cfg.GUI` into `updateGUIAuthentication`.

Reply:
In this special case, nothing else is running that might modify the config in parallel.  But in the general case, the `GUI()` method makes a copy while holding the lock, to make sure it's an atomic snapshot.  Should we simplify here based on the single-thread assumption?

Reply:
The `cfg *config.Configuration` you get as an argument is already a copy - that would be really bad if that was a pointer to the actual config object in use.